File: src/cpp/trapecioPool.cpp
// src/trapecio.cpp
#include <iostream>
#include <iomanip>
#include <limits>
#include <thread>
#include "CalculadorIntegralPool.h"
#include "ThreadPool.h"

int main()
{
    // Condiciones iniciales.
    const long double a = 2.0L;
    const long double b = 20.0L;

    // Tolerancia para la convergencia
    const long double epsilon = 1e-12L;

    const int num_hilos = std::thread::hardware_concurrency();

    ThreadPool pool(num_hilos);

    std::cout << "Usando pool de " << num_hilos << " hilos." << std::endl;

    long long n = 1; // Solo 1 trapecio inicialmente.
    long double resultadoAnterior = 0.0L;
    long double resultadoActual;

    std::cout << "Iniciando cálculo de la integral..." << std::endl;
    // Para mejorar la presentación de los números decimales.
    std::cout << std::fixed << std::setprecision(12);

    do
    {
        resultadoAnterior = resultadoActual;

        CalculadorIntegral calculador(a, b, n, pool);
        resultadoActual = calculador.calcular();

        std::cout << "Trapecios: " << std::setw(12) << std::left << n
                  << "| Área: " << resultadoActual << std::endl;

        // Limite de seguridad
        if (n > std::numeric_limits<long long>::max() / 2)
        {
            std::cout << "\nLimite de trapecios alcanzado. Deteniendo el calculo..." << std::endl;
            break;
        }
        n *= 2; // Duplicar numero de trapecios

        // La nueva condicion comprueba que la diferencia absoluta es mayor que la tolerancia
    } while (std::abs(resultadoActual - resultadoAnterior) > epsilon);

    std::cout << "\nConvergencia alcanzada." << std::endl;
    std::cout << "El valor aproximado de la integral es: " << resultadoActual << std::endl;
    std::cout << "Número de trapecios utilizado: " << n << std::endl;

    return 0;
}

File: src/java/TrapecioPool.java
import java.math.BigDecimal;
import java.math.MathContext;
import java.math.RoundingMode;

public class TrapecioPool {
    private static final MathContext MC = new MathContext(100);
    
    public static void main(String[] args) {
        final BigDecimal a = new BigDecimal("2.0");
        final BigDecimal b = new BigDecimal("20.0");
        final BigDecimal epsilon = new BigDecimal("1e-40"); // Tolerancia más estricta
        
        final int numHilos = Runtime.getRuntime().availableProcessors();
        ThreadPool pool = new ThreadPool(numHilos);
        
        System.out.println("=== MÉTODO DEL TRAPECIO - POOL DE THREADS ===");
        System.out.println("Usando pool de " + numHilos + " hilos.");
        System.out.println("Función: f(x) = 2x² + 3x + 0.5");
        System.out.println("Límites: [" + a + ", " + b + "]");
        System.out.println("Tolerancia: " + epsilon);
        System.out.println("\nIniciando cálculo...");
        System.out.println("Trapecios         | Área Aproximada");
        System.out.println("------------------|----------------------------------------");
        
        long n = 1;
        BigDecimal resultadoAnterior = BigDecimal.ZERO;
        BigDecimal resultadoActual;
        int iteracion = 0;
        final int MAX_ITERACIONES = 40; // Más iteraciones
        int iteracionesSinCambio = 0;
        final int MAX_ITERACIONES_SIN_CAMBIO = 3;
        
        try {
            // Primera iteración
            CalculadorIntegralPool calculador = new CalculadorIntegralPool(a, b, n, pool);
            resultadoActual = calculador.calcular();
            System.out.printf("%-16d | %s%n", n, formatDecimal(resultadoActual));
            n *= 2;
            iteracion++;
            
            do {
                resultadoAnterior = resultadoActual;
                
                calculador = new CalculadorIntegralPool(a, b, n, pool);
                resultadoActual = calculador.calcular();
                
                System.out.printf("%-16d | %s%n", n, formatDecimal(resultadoActual));
                
                // Calcular diferencia
                BigDecimal diferencia = resultadoActual.subtract(resultadoAnterior, MC).abs();
                
                iteracion++;
                
                // Verificar si hay cambio significativo
                if (diferencia.compareTo(epsilon) <= 0) {
                    iteracionesSinCambio++;
                    if (iteracionesSinCambio >= MAX_ITERACIONES_SIN_CAMBIO) {
                        System.out.println("\n Convergencia completa alcanzada (sin cambios significativos).");
                        break;
                    }
                } else {
                    iteracionesSinCambio = 0; // Reset counter
                }
                
                if (iteracion >= MAX_ITERACIONES) {
                    System.out.println("\nLímite máximo de iteraciones alcanzado.");
                    break;
                }
                
                n *= 2;
                
                // Protección por rendimiento
                if (n > 100_000_000L) {
                    System.out.println("\n Límite de trapecios alcanzado por rendimiento.");
                    break;
                }
                
            } while (true);
            
            System.out.println("\n--- RESULTADO FINAL ---");
            System.out.println("Valor de la integral: " + formatDecimal(resultadoActual));
            System.out.println("Trapecios utilizados: " + n);
            System.out.println("Iteraciones totales: " + iteracion);
            
        } finally {
            pool.shutdown();
        }
    }
    
    private static String formatDecimal(BigDecimal value) {
        String plain = value.toPlainString();
        
        // Asegurar que siempre muestre decimales
        if (!plain.contains(".")) {
            plain = plain + ".0";
        }
        
        // Limitar a 60 decimales máximo para mejor visualización
        if (plain.contains(".")) {
            String[] parts = plain.split("\\.");
            if (parts.length == 2 && parts[1].length() > 60) {
                plain = parts[0] + "." + parts[1].substring(0, 60);
            }
        }
        
        return plain;
    }
}